tensorflow/contrib/gdr/gdr_memory_manager.cc

```diff
@@ -390,92 +358,74 @@ void GdrMemoryManager::TransportOptionsFromTensor(
     ::google::protobuf::Any* mutable_transport_options, const Tensor& tensor,
     Device* device, DeviceContext* device_context, bool on_host,
     StatusCallback done) {
+  ibv_mr* mr = FindMemoryRegion(&tensor);
   auto buffer = DMAHelper::buffer(&tensor);
-  void* addr = buffer->data();
-  size_t length = buffer->size();
-  if (length == 0) {
-    done(errors::Unavailable("Cannot register tensor buffer of size 0"));
-    return;
-  }
-
-  ibv_mr* mr = FindMemoryRegion(addr, length);
 
 #if GOOGLE_CUDA
   if (device->tensorflow_gpu_device_info() && !on_host) {
     Allocator* alloc = GPUProcessState::singleton()->GetCUDAHostAllocator(0);
     Tensor* host_copy = new Tensor(alloc, tensor.dtype(), tensor.shape());
-    GPUUtil::CopyGPUTensorToCPU(
-        device, device_context, &tensor, host_copy,
-        [done, host_copy, mutable_transport_options, this](const Status& s) {
-          if (!s.ok()) {
-            done(s);
-            delete host_copy;
-            return;
-          }
-          auto buffer = DMAHelper::buffer(host_copy);
-          void* addr = buffer->data();
-          size_t length = buffer->size();
-          ibv_mr* mr = FindMemoryRegion(addr, length);
-
-          if (mr == nullptr) {
-            done(errors::Unavailable("Cannot find pinned memory region"));
-            delete host_copy;
-            return;
-          }
-
-          buffer->Ref();
-          TensorKey tensor_key = next_key_++;
-          {
-            mutex_lock l(server_mu_);
-            tensor_buffers_.insert(std::make_pair(tensor_key, buffer));
-          }
+    ibv_mr* mr = FindMemoryRegion(host_copy);
+    auto buffer = DMAHelper::buffer(host_copy);
+    if (mr == nullptr) {
+      done(errors::Unavailable("Cannot find pinned memory region"));
+      delete host_copy;
+      return;
+    }
 
-          uint64_t checksum = 0;
-          if (VLOG_IS_ON(2)) {
-            checksum = GPUUtil::Checksum(*host_copy);
-          }
+    buffer->Ref();
+    TensorKey tensor_key = next_key_++;
+    {
+      mutex_lock l(buf_mu_);
+      tensor_buffers_.insert(std::make_pair(tensor_key, buffer));
+    }
 
-          RemoteMemoryRegion remote_mr;
-          remote_mr.set_host(host_);
-          remote_mr.set_port(port_);
-          remote_mr.set_addr(reinterpret_cast<uint64_t>(addr));
-          remote_mr.set_rkey(mr->rkey);
-          remote_mr.set_tensor_key(tensor_key);
-          remote_mr.set_checksum(checksum);
-          mutable_transport_options->PackFrom(remote_mr);
+    uint64_t checksum = 0;
+    if (VLOG_IS_ON(2)) {
+      checksum = GPUUtil::Checksum(device, device_context, tensor);
```

Comment:
Nit.  GPUUtil::Checksum actually copies tensor to tmp CPU buffer then computes checksum there, since we don't (yet) have a GPU native checksum function.  So you could compute this on host_copy instead.

Reply:
I have reverted the changes in later commits as it is helpful to do an end-to-end checksum, especially when the tensor to be checked is read by NIC from GPU (aka GPUDirect RDMA). The checksum will only be used to debug, so I think it should be fine here.

Reply:
For simplicity, I will just leave it as-is. Since we only do checksum when debugging, it is better to do an end-to-end checksum for the final tensor value, instead of the intermediate copy. It saves a bigger callback for GPUUtil::CopyGPUTensorToCPU below.

Reply:
Resolved. Thanks for catching this!